// @ng-toolkit/pwa/bundles/ng-toolkit-pwa.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('rxjs'), require('@angular/core'), require('@angular/service-worker')) :
    typeof define === 'function' && define.amd ? define('@ng-toolkit/pwa', ['exports', 'rxjs', '@angular/core', '@angular/service-worker'], factory) :
    (factory((global['ng-toolkit'] = global['ng-toolkit'] || {}, global['ng-toolkit'].pwa = {}),global.rxjs,global.ng.core,global.ng['service-worker']));
}(this, (function (exports,rxjs,core,serviceWorker) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var SwUpdateServerMock = /** @class */ (function () {
        function SwUpdateServerMock() {
            this.available = new rxjs.Subject();
            this.activated = new rxjs.Subject();
            this.isEnabled = false;
        }
        /**
         * @return {?}
         */
        SwUpdateServerMock.prototype.checkForUpdate = /**
         * @return {?}
         */
            function () {
                return new Promise(function (resolve) { return resolve(); });
            };
        /**
         * @return {?}
         */
        SwUpdateServerMock.prototype.activateUpdate = /**
         * @return {?}
         */
            function () {
                return new Promise(function (resolve) { return resolve(); });
            };
        return SwUpdateServerMock;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var SwPushServerMock = /** @class */ (function () {
        function SwPushServerMock() {
        }
        /**
         * @param {?} options
         * @return {?}
         */
        SwPushServerMock.prototype.requestSubscription = /**
         * @param {?} options
         * @return {?}
         */
            function (options) {
                console.log("requested subscription with options: " + options);
                return new Promise(function (resolve) { return resolve(); });
            };
        /**
         * @return {?}
         */
        SwPushServerMock.prototype.unsubscribe = /**
         * @return {?}
         */
            function () {
                return new Promise(function (resolve) { return resolve(); });
            };
        return SwPushServerMock;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */
    var NgtPwaMockModule = /** @class */ (function () {
        function NgtPwaMockModule() {
        }
        NgtPwaMockModule.decorators = [
            { type: core.NgModule, args: [{
                        providers: [
                            { provide: serviceWorker.SwUpdate, useClass: SwUpdateServerMock },
                            { provide: serviceWorker.SwPush, useClass: SwPushServerMock }
                        ]
                    },] },
        ];
        return NgtPwaMockModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
     */

    exports.NgtPwaMockModule = NgtPwaMockModule;
    exports.SwPushServerMock = SwPushServerMock;
    exports.SwUpdateServerMock = SwUpdateServerMock;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
