// @ngx-share/core/esm2015/lib/share.models.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from '@angular/core';
/** @type {?} */
export const SHARE_BUTTONS_CONFIG = new InjectionToken('shareButtonsConfig');
/**
 * Share buttons config
 * @record
 */
export function ShareButtonsConfig() { }
if (false) {
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.prop;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.theme;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.include;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.exclude;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.size;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.url;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.title;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.description;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.image;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.tags;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.twitterAccount;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.autoSetMeta;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.gaTracking;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.windowWidth;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.windowHeight;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.moreButtonIcon;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.lessButtonIcon;
    /** @type {?|undefined} */
    ShareButtonsConfig.prototype.debug;
}
/**
 * OG meta tags
 * @record
 */
export function ShareMetaTags() { }
if (false) {
    /** @type {?|undefined} */
    ShareMetaTags.prototype.url;
    /** @type {?|undefined} */
    ShareMetaTags.prototype.title;
    /** @type {?|undefined} */
    ShareMetaTags.prototype.description;
    /** @type {?|undefined} */
    ShareMetaTags.prototype.image;
    /** @type {?|undefined} */
    ShareMetaTags.prototype.tags;
    /** @type {?|undefined} */
    ShareMetaTags.prototype.via;
}
/**
 * Share buttons collection
 * @record
 */
export function IShareButtons() { }
if (false) {
    /** @type {?|undefined} */
    IShareButtons.prototype.facebook;
    /** @type {?|undefined} */
    IShareButtons.prototype.twitter;
    /** @type {?|undefined} */
    IShareButtons.prototype.linkedin;
    /** @type {?|undefined} */
    IShareButtons.prototype.google;
    /** @type {?|undefined} */
    IShareButtons.prototype.tumblr;
    /** @type {?|undefined} */
    IShareButtons.prototype.mix;
    /** @type {?|undefined} */
    IShareButtons.prototype.pinterest;
    /** @type {?|undefined} */
    IShareButtons.prototype.reddit;
    /** @type {?|undefined} */
    IShareButtons.prototype.vk;
    /** @type {?|undefined} */
    IShareButtons.prototype.telegram;
    /** @type {?|undefined} */
    IShareButtons.prototype.messenger;
    /** @type {?|undefined} */
    IShareButtons.prototype.whatsapp;
    /** @type {?|undefined} */
    IShareButtons.prototype.xing;
    /** @type {?|undefined} */
    IShareButtons.prototype.line;
    /** @type {?|undefined} */
    IShareButtons.prototype.sms;
    /** @type {?|undefined} */
    IShareButtons.prototype.email;
    /** @type {?|undefined} */
    IShareButtons.prototype.print;
    /** @type {?|undefined} */
    IShareButtons.prototype.copy;
}
/**
 * Share button properties
 * @record
 */
export function IShareButton() { }
if (false) {
    /**
     * A function that returns a new share button object
     * @type {?|undefined}
     */
    IShareButton.prototype.create;
    /**
     * Share button text
     * @type {?|undefined}
     */
    IShareButton.prototype.text;
    /**
     * Share button icon (FontAwesome)
     * @type {?|undefined}
     */
    IShareButton.prototype.icon;
    /**
     * Share button color
     * @type {?|undefined}
     */
    IShareButton.prototype.color;
    /**
     * Share button aria label attribute
     * @type {?|undefined}
     */
    IShareButton.prototype.ariaLabel;
    /**
     * Used to attach more properties for certain buttons
     * @type {?|undefined}
     */
    IShareButton.prototype.extra;
}
