// @ngx-share/core/esm2015/lib/buttons/twitter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ShareButtonBase } from './base';
export class TwitterButton extends ShareButtonBase {
    /**
     * @param {?} _props
     * @param {?} _url
     * @param {?} _http
     * @param {?} _platform
     * @param {?} _document
     * @param {?} _windowSize
     * @param {?} _disableButtonClick
     * @param {?} _logger
     */
    constructor(_props, _url, _http, _platform, _document, _windowSize, _disableButtonClick, _logger) {
        super(_props, _url, _http, _platform, _document, _windowSize, _disableButtonClick, _logger);
        this._props = _props;
        this._url = _url;
        this._http = _http;
        this._platform = _platform;
        this._document = _document;
        this._windowSize = _windowSize;
        this._disableButtonClick = _disableButtonClick;
        this._logger = _logger;
        this._supportedMetaTags = {
            url: 'url',
            description: 'text',
            tags: 'hashtags',
            via: 'via'
        };
    }
    /**
     * @return {?}
     */
    get desktop() {
        return `https://twitter.com/intent/tweet?`;
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._supportedMetaTags;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._props;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._url;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._http;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._platform;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._document;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._windowSize;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._disableButtonClick;
    /**
     * @type {?}
     * @protected
     */
    TwitterButton.prototype._logger;
}
